docs/ql-style-guide.md

```diff
@@ -0,0 +1,434 @@
+# QL Style Guide
+
+## Introduction
+
+This document describes how to format the QL code in this repository, covering aspects such as layout, white-space, naming and documentation. Adhering to consistent standards makes code easier to read and maintain. Of course, these are only guidelines, and can be overridden as the need arises on a case-by-case basis. Where existing code deviates from these guidelines, prefer consistency with the surrounding code.
+
+Words in *italic* are defined in the [Glossary](#glossary).
+
+## Indentation
+1. *Always* use 2 spaces for indentation.
+1. *Always* indent:
+   - the *body* of a module, newtype, class or predicate,
+   - the second and subsequent lines after a line break used to split a long line,
+   - the *body* of a `from`, `where` or `select` clause where it spans multiple lines,
+   - the *body* of a *quantifier* that spans multiple lines.
+
+
+Examples:
+
+```ql
+module Helpers {
+  /** ... */
+  class X ... {
+    /** ... */
+    int getNumberOfChildren () {
+      result = count(int child |
+        exists(this.getChild(child))
+      )
+    }
+  }
+}
+```
+
+```ql
+from Call c, string reason
+where isDeprecated(c, reason)
+select c, "This call to '$@' is deprecated because " + reason + ".",
+  c.getTarget(), c.getTarget().getName()
+```
+
+## Line breaks
+1. Use UNIX line endings.
+1. Lines *must not* exceed 100 characters.
+1. Long lines *should* be split with a line break, and the following lines *must* be indented one level until the next "regular" line break.
+1. There *should* be a single blank line:
+   - between the file documentation and the first `import`,
+   - before each declaration, except for the first declaration in a *body*,
+   - before the `from`-`where`-`select` section in a query file.
+1. *Avoid* two or more adjacent blank lines.
+1. There *must* be a new line after the *annotations* `cached`, `deprecated`, `pragma`, `language` and `bindingset`.
+1. There *should not* be additional blank lines within a predicate.
+1. There *may* be a new line:
+   - immediately after the `from`, `where` or `select` keywords in a query,
+   - immediately after `if`, `then`, or `else` keywords. The `then` and `else` parts *should* be consistent.
+1. *Avoid* other line breaks in declarations, other than to break long lines.
+1. When operands of *binary operators* span two lines, the operator *should* be placed at the end of the first line.
+
+Examples:
+
+```ql
+cached
+private int getNumberOfParameters() {
+  ...
+}
+```
+
+```ql
+predicate methodStats(string qualifiedName, string name,
+  int numberOfParameters, int numberOfStatements, int numberOfExpressions,
+  int linesOfCode, int nestingDepth, int numberOfBranches) {

```

Comment:
As far as I understood there wasn't much discussion around the different options for long parameter list, other than the fact that we need to address it.  Indenting one level is as I see it one of the bad options, since it'll match the indentation of the body.  So we should probably either use double-indentation or place the `) {` on its own line.  I'm not sure which of those two options is the nicest...

Reply:
This wouldn't be an issue if we just put the open curly on its own line where it belongs;) Assuming that option's off the table, though, I prefer the double indentation.

Reply:
I also prefer putting braces on their own line. However it seems we're outvoted. I've decided to defer discussion on this  -- we should avoid long list of arguments anyway.

My own preference is `) {` on its own line.